ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java

```diff
@@ -0,0 +1,596 @@
+  private byte[] userMetadata;
+  private byte[] putContent;
+
+  private final GetTestRequestRegistrationCallbackImpl requestRegistrationCallback =
+      new GetTestRequestRegistrationCallbackImpl();
+
+  private class GetTestRequestRegistrationCallbackImpl implements RequestRegistrationCallback<GetOperation> {
+    private List<RequestInfo> requestListToFill;
+
+    @Override
+    public void registerRequestToSend(GetOperation getOperation, RequestInfo requestInfo) {
+      requestListToFill.add(requestInfo);
+      correlationIdToGetOperation.put(((RequestOrResponse) requestInfo.getRequest()).getCorrelationId(), getOperation);
+    }
+  }
+
+  private final AtomicInteger operationsCount = new AtomicInteger(0);
+  private final OperationCompleteCallback operationCompleteCallback = new OperationCompleteCallback(operationsCount);
+
+  public GetBlobOperationTest() {
+    // Defaults. Tests should override these as appropriate.
+    maxChunkSize = random.nextInt(1024 * 1024) + 1;
+    blobSize = maxChunkSize * random.nextInt(10) + random.nextInt(maxChunkSize - 1) + 1;
+  }
+
+  @After
+  public void after() {
+    if (networkClient != null) {
+      networkClient.close();
+    }
+    Assert.assertEquals("All operations should have completed", 0, operationsCount.get());
+  }
+
+  /**
+   * Instantiate a router, perform a put, close the router. The blob that was put will be saved in the MockServer,
+   * and can be queried by the getBlob operations in the test.
+   */
+  private void doPut()
+      throws Exception {
+    requestParallelism = 3;
+    successTarget = 2;
+    mockSelectorState.set(MockSelectorState.Good);
+    VerifiableProperties vprops = new VerifiableProperties(getNonBlockingRouterProperties());
+    routerConfig = new RouterConfig(vprops);
+    mockClusterMap = new MockClusterMap();
+    routerMetrics = new NonBlockingRouterMetrics(mockClusterMap.getMetricRegistry());
+    mockServerLayout = new MockServerLayout(mockClusterMap);
+    replicasCount = mockClusterMap.getWritablePartitionIds().get(0).getReplicaIds().size();
+    responseHandler = new ResponseHandler(mockClusterMap);
+    networkClientFactory = new MockNetworkClientFactory(vprops, mockSelectorState, MAX_PORTS_PLAIN_TEXT, MAX_PORTS_SSL,
+        CHECKOUT_TIMEOUT_MS, mockServerLayout, time);
+    router =
+        new NonBlockingRouter(routerConfig, new NonBlockingRouterMetrics(new MetricRegistry()), networkClientFactory,
+            new LoggingNotificationSystem(), mockClusterMap, time);
+    blobProperties = new BlobProperties(blobSize, "serviceId", "memberId", "contentType", false, Utils.Infinite_Time);
+    userMetadata = new byte[BLOB_USER_METADATA_SIZE];
+    random.nextBytes(userMetadata);
+    putContent = new byte[blobSize];
+    random.nextBytes(putContent);
+    ReadableStreamChannel putChannel = new ByteBufferReadableStreamChannel(ByteBuffer.wrap(putContent));
+    blobIdStr = router.putBlob(blobProperties, userMetadata, putChannel).get();
```

Comment:
There is a tradeoff. The alternative is for all these methods to share state via return values (like what you suggest) or pass them as arguments. Which is not any better either.

The only value is if variables can be local to a method - in which case they should not be made members. Otherwise, if different methods are sharing the state, might as well make them members.

For this particular example, if `doPut()` could return the `blobIdStr`, but then certain tests will have to pass it down to helper methods. If the variable gets passed across 2-3 methods of the class, why not simply make it a member variable?